api/ai/new/priority_algorithm/priority_algorithm.py

```diff
@@ -0,0 +1,174 @@
+    mutate_robinhood_holistic,
+)
+from api.ai.new.priority_algorithm.priority_teamset import PriorityTeamSet, PriorityTeam
+from api.ai.new.weight_algorithm.weight_algorithm import WeightAlgorithm
+from api.models.student import Student
+from api.models.team import Team
+from api.models.team_set import TeamSet
+
+DEFAULT_PRIORITY_ALGORITHM_CONFIG = PriorityAlgorithmConfig(
+    MAX_KEEP=3,
+    MAX_SPREAD=3,
+    MAX_ITERATE=1500,
+    MAX_TIME=1,
+)
+
+
+class PriorityAlgorithm(Algorithm):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.algorithm_options: PriorityAlgorithmOptions = cast(
+            PriorityAlgorithmOptions, self.algorithm_options
+        )
+        if self.algorithm_config:
+            self.algorithm_config: PriorityAlgorithmConfig = cast(
+                PriorityAlgorithmConfig, self.algorithm_config
+            )
+        else:
+            self.algorithm_config = DEFAULT_PRIORITY_ALGORITHM_CONFIG
+
+        self.student_dict: Dict[int, Student] = {}
+
+    def generate_initial_team_set(
+        self,
+        students: List[Student],
+    ) -> PriorityTeamSet:
+        team_set = WeightAlgorithm(
+            # todo: implement proper conversion from priority algorithm options to weight algorithm options
+            algorithm_options=weight_options_from_priority_options(
+                self.algorithm_options
+            ),
+            team_generation_options=self.team_generation_options,
+        ).generate(students)
+
+        return PriorityTeamSet(
+            priority_teams=[
+                PriorityTeam(team=team, student_ids=[s.id for s in team.students])
+                for team in team_set.teams
+            ]
+        )
+
+    def generate(self, students: List[Student]) -> TeamSet:
+        self.student_dict = create_student_dict(students)
+
+        start_time = time.time()
+        iteration = 0
+        team_sets = [self.generate_initial_team_set(students)]
+
+        while (
+            time.time() - start_time < self.algorithm_config.MAX_TIME
+            and iteration < self.algorithm_config.MAX_ITERATE
+        ):
+            new_team_sets: List[PriorityTeamSet] = []
+            for team_set in team_sets:
+                new_team_sets += self.mutate(team_set)
+            team_sets = new_team_sets + team_sets
+            team_sets = sorted(
+                team_sets,
+                key=lambda ts: ts.calculate_score(
+                    self.algorithm_options.priorities, self.student_dict
+                ),
+                reverse=True,
+            )
+            team_sets = team_sets[: self.algorithm_config.MAX_KEEP]
+            iteration += 1
+
+        # the first team set is the "best" one
+        return self._unpack_priority_team_set(team_sets[0])
+
+    def mutate(self, team_set: PriorityTeamSet) -> List[PriorityTeamSet]:
+        """
+        Mutate a single teamset into child teamsets
+        """
+        algorithm = 1
```

Comment:
I think this should be a part of the `PriorityAlgorithmConfig`

Reply:
Oh yeah I gotta do the full refactor for it though